mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp

```diff
@@ -125,34 +134,60 @@ struct ToQuantumComputationPattern final : mlir::OpRewritePattern<AllocOp> {
    * @param op The operation to convert.
    * @param currentQubitVariables The list of previously defined qubit
    * variables.
+   *
+   * return if the operation was successfully handled
    */
-  void handleUnitaryOp(UnitaryInterface& op,
+  bool handleUnitaryOp(UnitaryInterface op,
                        std::vector<mlir::Value>& currentQubitVariables) const {
+
+    // Add the operation to the QuantumComputation.
+    qc::OpType opType = qc::OpType::H; // init placeholder-"H" overwritten next
+    if (llvm::isa<HOp>(op)) {
+      opType = qc::OpType::H;
+    } else if (llvm::isa<XOp>(op)) {
+      opType = qc::OpType::X;
+    } else { // TODO: support for more operations
+      throw std::runtime_error("Unsupported operation type!");
+    }
+
     const auto in = op.getInQubits()[0];
     const auto ctrlIns = op.getCtrlQubits();
     const auto outs = op.getOutQubits();
 
-    // Get the qubit index of every control qubit.
-    std::vector<size_t> ctrlInsIndices(ctrlIns.size());
-    std::transform(ctrlIns.begin(), ctrlIns.end(), ctrlInsIndices.begin(),
-                   [&currentQubitVariables](const mlir::Value val) {
-                     return findQubitIndex(val, currentQubitVariables);
-                   });
+    try {
+      // Get the qubit index of every control qubit.
+      std::vector<size_t> ctrlInsIndices(ctrlIns.size());
+      std::transform(ctrlIns.begin(), ctrlIns.end(), ctrlInsIndices.begin(),
+                     [&currentQubitVariables](const mlir::Value val) {
+                       return findQubitIndex(val, currentQubitVariables);
+                     });
 
-    // Get the qubit index of the target qubit.
-    const size_t targetIndex = findQubitIndex(in, currentQubitVariables);
+      // Get the qubit index of the target qubit.
+      const size_t targetIndex = findQubitIndex(in, currentQubitVariables);
 
-    // Update `currentQubitVariables` with the new qubit values.
-    for (size_t i = 0; i < ctrlInsIndices.size(); i++) {
-      currentQubitVariables[ctrlInsIndices[i]] = outs[i + 1];
+      // Update `currentQubitVariables` with the new qubit values.
+      for (size_t i = 0; i < ctrlInsIndices.size(); i++) {
+        currentQubitVariables[ctrlInsIndices[i]] = outs[i + 1];
+      }
+      currentQubitVariables[targetIndex] = outs[0];
+
+      // Add the operation to the QuantumComputation.
+      auto operation = qc::StandardOperation(
+          qc::Controls{ctrlInsIndices.cbegin(), ctrlInsIndices.cend()},
+          targetIndex, opType);
+      circuit.push_back(operation);
```

Comment:
```suggestion
      circuit.emplace_back<StandardOperation(qc::Controls{ctrlInsIndices.cbegin(), ctrlInsIndices.cend()},
          targetIndex, opType);
```
This should work and is more efficient as it constructs the operation inplace.